File: JavaScript/OCP_demo.js
class IDiscount {
    getDiscountedPrice(amount) {
        throw new Error("Not implemented");
    }
}

class Festival extends IDiscount {
    constructor() {
        super();
        this.discount = 0.25;
    }
    getDiscountedPrice(amount) {
        return amount * this.discount;
    }
}

class Holi extends IDiscount {
    constructor() {
        super();
        this.discount = 0.25;
    }
    getDiscountedPrice(amount) {
        return amount * this.discount;
    }
}

class PaymentService {
    constructor(discount) {
        this.discount = discount;
    }
    pay(amount) {
        const final_ = this.discount.getDiscountedPrice(amount);
        console.log(final_);
    }
}

class OCP_Demo {
    static main(args) {
        const service = new PaymentService(new Festival()); // solution is Factory Method design Pattern
    }
}

OCP_Demo.main([]);

File: Java/OCP_demo.java

interface IDiscount{
    float getDiscountedPrice(int amount);
}

class Festival implements IDiscount{
    float discount = 0.25f;
    @Override
    public float getDiscountedPrice(int amount) {
        return amount*discount;
    }

}

class Holi implements IDiscount{
    float discount = 0.25f;
    @Override
    public float getDiscountedPrice(int amount) {
        return amount*discount;
    }
}

class PaymentService{
    IDiscount discount;

    public PaymentService(IDiscount discount){
        this.discount = discount;
    }
    void pay(int amount){
        float final_ = discount.getDiscountedPrice(amount);
        System.out.println(final_);
    }
}

public class OCP_Demo {
    public static void main(String[] args) {

        PaymentService service = new PaymentService(new Festival());    //solution is Factory Method design Pattern
    }
}
